apps/passport-client/src/validateState.ts

```diff
@@ -0,0 +1,174 @@
+  const commitmentFromSelfField = state?.self?.commitment;
+  const commitmentFromIdentityField = state?.identity?.commitment?.toString();
+
+  if (commitmentOfIdentityPCDInCollection !== commitmentFromSelfField) {
+    validationErrors.push(
+      `commitment of identity pcd in collection (${commitmentOfIdentityPCDInCollection})` +
+        ` does not match commitment in 'self' field of app state (${commitmentFromSelfField})`
+    );
+  }
+
+  if (commitmentFromSelfField !== commitmentFromIdentityField) {
+    validationErrors.push(
+      `commitment in 'self' field of app state (${commitmentFromSelfField})` +
+        ` does not match commitment of 'identity' field of app state (${commitmentFromIdentityField})`
+    );
+  }
+
+  if (commitmentFromIdentityField !== commitmentOfIdentityPCDInCollection) {
+    validationErrors.push(
+      `commitment of 'identity' field of app state (${commitmentFromIdentityField})` +
+        ` does not match commitment of identity pcd in collection (${commitmentOfIdentityPCDInCollection})`
+    );
+  }
+
+  return validationErrors;
+}
+
+/**
+ * Validates a {@link PCDCollection} by checking its contents. Does verify that the collection
+ * contains PCDs that are consistent with the rest of the application state. Returns a list of
+ * strings representing individual errors. If there are no errors, returns an empty array.
+ */
+export function validatePCDCollection(pcds?: PCDCollection): string[] {
+  const validationErrors: string[] = [];
+
+  if (!pcds) {
+    validationErrors.push("pcd collection is absent");
+  }
+
+  if (pcds.size() === 0) {
+    validationErrors.push("pcd collection is empty");
+  }
+
+  const identityPCDFromCollection = pcds.getPCDsByType(
+    SemaphoreIdentityPCDPackage.name
+  )[0] as SemaphoreIdentityPCD | undefined;
+
+  if (!identityPCDFromCollection) {
+    validationErrors.push("pcd collection does not contain an identity pcd");
+  }
+
+  return validationErrors;
+}
+
+export function validateUpload(user?: User, pcds?: PCDCollection): string[] {
+  const validationErrors: string[] = [];
+
+  if (!user) {
+    validationErrors.push(`upload must include a user`);
+  }
+
+  if (!pcds) {
+    validationErrors.push(`upload must include a pcd collection`);
+  }
+
+  const identityPCDFromCollection = pcds.getPCDsByType(
+    SemaphoreIdentityPCDPackage.name
+  )[0] as SemaphoreIdentityPCD | undefined;
+  const commitmentFromPCDCollection =
+    identityPCDFromCollection?.claim?.identity?.commitment?.toString();
+
+  if (user?.commitment !== commitmentFromPCDCollection) {
+    validationErrors.push(
+      "user commitment does not equal to commitment of identity pcd in pcd collection"
+    );
+  }
+
+  return validationErrors;
+}
+
+/**
+ * Validates whether a user returned by the Zupass server API is consistent
+ * with the local application state representation. Returns errors as strings,
+ * and returns an empty array if the two are not inconsistent. Does not validate
```

Comment:
Why does this not validate AppState in its entirety?

Reply:
I was not sure whether the functions I created here should have overlapping responsibilities, or whether each one should handle its own situation. This function is intended to be called only from `finishAccountCreation`, from which I extracted the validation logic and placed in this function. I could see an argument for invoking `validateState` from inside this function, or for, as I mentioned in an earlier comment, creating a version of `validateState` that takes as parameters fields that the caller of `validateState` wants to update in `appState`. How do you think this can be best structured?

Reply:
I think you may as well validate everything you can.  Given this function has access to a full AppState it should be possible to validate it.  From another angle, it might be good to do that full validation on all of the paths which get a user into a logged-in state: initial load, finishAccountCreation, loadAfterLogin.

Separate but more risky, but what I think you're hinting toward with the idea of validating state updates: There could be validation inside of `update()`, which has access to full AppState and the changes being made to it.  Validating there would be a very solid way of ensuring invariants never get broken at that level, though I'm not sure how easy it would be to respond or signal to the user from that level.  Unlike specific scenarios (upload, download, login) it's not obvious what the scenario is or where the data is coming from.  Maybe there could be validation at that level which only logs without changing behavior?

Reply:
performing validation inside of `update` is actually a really good idea.

Reply:
So good that it sent me on a whole tangent in #1359